tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/quantization/QuantizationTests.java

```diff
@@ -0,0 +1,891 @@
+        int outputDim = 2048; // Default output dimension (rows)
+
+        FloatArray input = new FloatArray(inputDim);
+        FloatArray weights = new FloatArray(inputDim * outputDim);
+        FloatArray output = new FloatArray(outputDim);
+        FloatArray outputSeq = new FloatArray(outputDim);
+        fillRandomData(input, -1.0f, 1.0f, random);
+        fillRandomData(weights, -0.1f, 0.1f, random);
+
+        // already quantized
+        Int8Array w_quant = new Int8Array(weights.getSize());
+        FloatArray w_scale = new FloatArray(1);
+        quantizeFloatArray(weights, w_quant, w_scale);
+
+        FloatArray x_scale = new FloatArray(1);
+        int maxNumGroups = (inputDim + local_workgroup_size - 1) / local_workgroup_size;
+        FloatArray x_max = new FloatArray(maxNumGroups);
+        FloatArray inv_scale = new FloatArray(1);
+        Int8Array x_quant = new Int8Array(input.getSize());
+
+        WorkerGrid1D workerQuant = new WorkerGrid1D(inputDim);
+        WorkerGrid1D workerDp4a = new WorkerGrid1D(local_workgroup_size * inputDim);
+
+        GridScheduler schedulerDp4a = new GridScheduler();
+        schedulerDp4a.addWorkerGrid("s0_quant_kc.scales", workerQuant);
+        schedulerDp4a.addWorkerGrid("s0_quant_kc.quantize", workerQuant);
+        schedulerDp4a.addWorkerGrid("s0_quant_kc.dp4amatvec", workerDp4a);
+
+        workerQuant.setLocalWork(local_workgroup_size, 1, 1);
+        workerDp4a.setLocalWork(local_workgroup_size, 1, 1);
+
+        TaskGraph taskGraph = new TaskGraph("s0_quant_kc")
+                .transferToDevice(DataTransferMode.FIRST_EXECUTION, input, x_quant, x_scale, x_max, inv_scale, w_quant, w_scale)
+                .task("scales", QuantizationTests::reductionCalculateMax, new KernelContext(), x_max, input, x_scale, inv_scale, local_workgroup_size, inputDim)
+                .task("quantize", QuantizationTests::quantizeKernelContext,new KernelContext(), input, inv_scale, x_quant)
+                .task("dp4amatvec", QuantizationTests::matrixVectorGenericPacked, new KernelContext(), w_quant, x_quant,
+                        output, inputDim, outputDim, local_workgroup_size, w_scale, x_scale)
+                .transferToHost(DataTransferMode.EVERY_EXECUTION, output, x_quant, x_scale, inv_scale, x_max);
+
+        ImmutableTaskGraph immutabletaskGraph = taskGraph.snapshot();
+        try (TornadoExecutionPlan tornadoExecutor = new TornadoExecutionPlan(immutabletaskGraph)) {
+            tornadoExecutor.withGridScheduler(schedulerDp4a).execute();
+        }
+
+        matrixVectorSequential(outputSeq, weights, input, inputDim, outputDim);
+
+        for (int i = 0; i < output.getSize(); i++) {
+            assertEquals(outputSeq.get(i), output.get(i), 0.1f);
+        }
+
+    }
+
+    @Test
+    public void testMatrixVector4WayDP4AKernel() throws TornadoExecutionPlanException {
+        assertNotBackend(TornadoVMBackendType.OPENCL);
```

Comment:
why this and previous tests are exempted for OpenCL?

Also, can you please @mairooni add the unit-tests in the test entries within the `tornado-test` to be executed by Jenkins?

Reply:
thanks, I missed the `tornado-test` file. sorry